main/src/ca/uwaterloo/flix/language/phase/Weeder.scala

```diff
@@ -2053,21 +2053,56 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
   /**
     * Weeds the given type parameters `tparams0`.
     */
-  private def visitTypeParams(tparams0: ParsedAst.TypeParams): WeededAst.TypeParams = tparams0 match {
-    case ParsedAst.TypeParams.Elided => WeededAst.TypeParams.Elided
+  private def visitTypeParams(tparams0: ParsedAst.TypeParams): Validation[WeededAst.TypeParams, WeederError] = tparams0 match {
+    case ParsedAst.TypeParams.Elided => WeededAst.TypeParams.Elided.toSuccess
     case ParsedAst.TypeParams.Explicit(tparams) =>
       val newTparams = tparams.map(visitTypeParam)
-      WeededAst.TypeParams.Explicit(newTparams)
+      val kindedTypeParams = newTparams.collect { case t: WeededAst.TypeParam.Kinded => t }
+      val unkindedTypeParams = newTparams.collect { case t: WeededAst.TypeParam.Unkinded => t }
+      (kindedTypeParams, unkindedTypeParams) match {
+          // Case 1: only unkinded type parameters
+        case (Nil, _ :: _) => WeededAst.TypeParams.Unkinded(unkindedTypeParams).toSuccess
+          // Case 2: only kinded type parameters
+        case (_ :: _, Nil) => WeededAst.TypeParams.Kinded(kindedTypeParams).toSuccess
+          // Case 3: some unkinded and some kinded
+        case (_ :: _, _ :: _) =>
+          val loc = mkSL(tparams.head.sp1, tparams.last.sp2)
+          WeederError.InconsistentTypeParameters(loc).toFailure
+          // Case 4: no type parameters: should be prevented by parser
+        case (Nil, Nil) => throw InternalCompilerException("Unexpected empty type parameters.")
```

Comment:
I would put the (Nil, Nil) case with the exception first and then I would replace the use of `_ :: _` with just `_`.

Reply:
I try to avoid using `_` as an "anything else" indicator, to ensure I'm not missing any cases.

Reply:
I see, but in this case it is more a question of emptyness vs. non-emptyness. You could also use those.

Anyway, I am fine with whatever.